src/aspire/nfft/cufinufft.py

```diff
@@ -0,0 +1,131 @@
+from ctypes import c_int
+
+import logging
+import numpy as np
+import pycuda.autoinit
+import pycuda.driver as cuda
+import pycuda.gpuarray as gpuarray
+from cufinufft import cufinufft
+
+from aspire.nfft import Plan
+from aspire.utils import ensure
+
+logger = logging.getLogger(__name__)
+
+class cuFINufftPlan(Plan):
+    def __init__(self, sz, fourier_pts, epsilon=1e-15, many=0, **kwargs):
+        """
+        A plan for non-uniform FFT in 2D or 3D.
+
+        :param sz: A tuple indicating the geometry of the signal
+        :param fourier_pts: The points in Fourier space where the Fourier transform is to be calculated,
+            arranged as a dimension-by-K array. These need to be in the range [-pi, pi] in each dimension.
+        :param epsilon: The desired precision of the NUFFT
+        :param many: Optional integer indicating if you would like to compute a batch of `many`
```

Comment:
I think I was worried `ntransforms=1` is ambiguous with the regular call in a functional/non-aethsetic way in the code, but I'd have to check. If it isn't then this would be simpler, yes.

Reply:
I can get the information by using the dimension from `sz` at Plan instantiation and then by inferring when we are dimension+1 that the additional axis is for the stack in the forward transform. Similarly, if we're stack x num_pts in the adjoint. 

We still need a many/ntransforms argument, because the signal is given at plan creation (when we need to pack ntranforms).  I added some assurances that these are coherent. Seems to work. It is in one single commit if there is a problem etc.

Reply:
I don't quite understand the ambiguity. You're talking about the `nufft3` and `anufft3` functions?

If we want to infer `ntransforms`, this has to be done during the call to `transform` or `adjoint`, so that won't work (since we want to instantiate the plan in `__init__` and need to know `ntransforms` there). For the plan, it makes sense to specify it, I think.

Reply:
That isn't the ambiguity.

Consider the shape of things for `ntransforms=1`.  You might have (1,n,n)  or (n,n) for the transform. These have implications in what they would return if the calling code expects them to be generic (admitting stack of 1, then return stack of 1).  If we don't admit stack of one I would have to mess with a bunch of code that doesn't relate to cufinufft in this PR (yes we did that stack of one about half the time).  I wrote a change that will work with our codebase... its not clearer to me, but it might be clearer to others.

Reply:
(I totally agree with what your saying about when we need `ntransforms`, I guess I didn't do a great job explaining it in my second sentence)

Reply:
We should definitely be admitting a stack of one when `ntransforms == 1`. Could this not be tracked during the `forward`/`adjoint` calls? That is, if we get `(n, n)` as input, we output `(m,)` and if we get `(1, n, n)` as input, we output `(1, m)`? So we respect the shape of the input. If `ntransforms > 0`, we of course require inputs to be `(ntransforms, n, n)` and return `(ntransforms, m)`.

Reply:
Briefly, that's what the code does more or less. Though the shapes are "funny" because I was trying to match what was done before me (trying not disturb all the ordering things in this PR just to add this feature). It was pushed yesterday morning. I think I distracted us with my discussion comments :).

details...
In what I originally pushed for review I called the `(n,n)` case the default of `many=0`.  This was so `many=1` doesn't mean two different things depending on other inputs. It also meant that I had a simple way to set a flag in `*plan.__init__` that would control shape logic later (below we have slightly different logic later in two places (forward/adjoint)). They both net the same effect of packing `ntransform` for the low level code and setting resultant shapes.

Currently, based on the change request, there is some logic whereby this `(n,n)` case is `ntransform=1` and in the `forward/adjoint` calls we use the  shape of signal array compared to shape of `sz`/`pts`. If it is the case we have `(1,n,n)` (well actually `(n,n,1)` sigh) then we know to shape the result as a stack. The branch logic is something like `len(sz) == signal.shape + 1`, of course with other conditions . Like I said, it is just a different way of doing the same thing. This is a classic corner case.  I find it distasteful to have ambiguous corner cases, but since we are wrapping these up in another layer, it's really no concern to me... we just duplicate similar logic in the wrapper as well to yield your auto-detection... all good.

I think a lot of the confusion in the discussion was me mixing up whether we were talking about `__init__.py` vs `cufinufft.py`. We've discussed at great length that have different information available to them.  My bad.

I'll rm `sz` in the wrapper function and calls. Otherwise are we looking good for a first pass? I know very well that this will need another, smaller pass after ordering changes (here and FI release), so I mainly want to make sure we're serviceable for a while and that you are happy with the API-ish layer (`anufft`/`nufft`). Thanks

Reply:
Ok good. Then we're fine for this PR.

Just for my own curiosity, you would rather have two cases: `ntransforms == 0`, which means an input of size `(n, n)` and `ntransforms == 1`, which means an input of size `(n, n, 1)`. That would make sense, but I see how that would touch more code than necessary at this point.

Reply:
Yes, as it was originally was intentional.  Same or slightly less code.  The detection at plan instantiation is not different for `transform` / `adjoint` in that approach, since it is just explicit (by arg 0) at the low level plan instantiation.  The only hurdle is a user wanting stack of one would have to prescribe an optional argument `==1`, but the user probably should not be using that layer directly anyway... it would be the wrapper, which we've done here now.

Both ways allow the `anufft/nufft` wrapper to work the same, same function signature, so it doesn't matter.. in the sense that this can always be changed in the future if there is a reason to do so. Since you will be living with this code long term, I think it's best to go with your preference.

Reply:
Got it.